libs/community/langchain_community/document_loaders/parsers/excel.py

```diff
@@ -0,0 +1,89 @@
+from typing import Iterator, Optional
+
+from langchain_core.documents import Document
+
+from langchain_community.document_loaders.base import BaseBlobParser
+from langchain_community.document_loaders.blob_loaders import Blob
+
+
+class ExcelParser(BaseBlobParser):
+    """Parse the Microsoft Excel documents from a blob.
+    - Uses Unstructured for parsing
+    - Each sheet is returned as a separate document.
+    - Metadata contains:
+      - source: from blob.source
+      - title: sheet name
+      - page: ordinal number of the sheet
+    - page_content contains Markdown:
+      - Elements identified as "Title" are prepended with `# `
+      - Table element's html representation is parsed using `htmltabletomd`
+      - All other elements are passed in "as is"
+    """
+
+    def lazy_parse(self, blob: Blob) -> Iterator[Document]:  # type: ignore[valid-type]
+        """Parse a Microsoft Excel document into the Document iterator.
+
+        Args:
+            blob: The blob to parse.
+
+        Returns: An iterator of Documents.
+
+        """
+        try:
+            from unstructured.partition.xlsx import partition_xlsx
```

Comment:
Thanks for this. How do you feel about using [langchain-unstructured](https://github.com/langchain-ai/langchain-unstructured/tree/main/libs/unstructured) here? Are there important differences in the resulting `Document` objects?

Reply:
@ccurme Actually in the meantime, I've written up [this](https://github.com/langchain-ai/langchain/pull/27749) wrapper that allows to turn a `DocumentLoader` into `Parser`. Excel parser is already implemented [here](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/document_loaders/excel.py) so I'd abandon this PR in favor of merging the `DocumentLoaderAsParser` wrapper together with some more changes to `UnstructuredExcelLoader` (or possibly `langchain-unstructured.document_loaders.UnstructuredLoader`). What's the current thinking? Is the implementation in community being deprecated? (In community there is implementation with inheritance `UnstructuredExcelLoader : UnstructuredFileLoader : UnstructuredBaseLoader : BaseLoader` where as in `langchain-unstructured` there is independent implementation only inheriting directly from `BaseLoader`.

Reply:
@ccurme I see that the [DocumentLoaderAsParser](https://github.com/langchain-ai/langchain/pull/27749) was assigned status _needs support_. I would appreciate if you could take a look at that one in context of being substitute for this PR. Thanks!

Reply:
@ccurme I'm thinking I might not have made clear enough about why having `BlobParser` rather than `DocumentLoader` is important. Please let me breifly explain:
- Document loaders function as `<whatever> -> documents`
- BlobParsers process `blob -> documents`.
- Many document loaders actually implement logic until they have a blob. Then the actual parsing is done by calling parser with similar name. As an example you can look at `AmazonTextractPDFLoader` and `AmazonTextractPDFParser`.
- Some document loaders may want to process numerous filetypes. An example is a `SharePointLoader` that is supposed to fetch all parsable files, parse them and produce documents. Until recently it processed only 3 file types `doc, docx, pdf` and ignored everything else. Recently I got [this pr](https://github.com/langchain-ai/langchain/pull/27716) merged that enables a user to pass in dict mapping file types (or mime types) to parsers. Now one can easily use `SharePointLoader` with any files where suitable parser is available.
- Unfortunatelly many document loaders do not define `BlobParser` but rather handle everything directly inside the loader class. `UnstructuredLoader` is one of those. 
- Originally I went ahead and implemented separate `ExcelParser` (this PR). However this duplicated some of the code. That could be mitigated by replacing the part of the code inside `UnstructuredExcelLoader` responsible for parsing the file with calling the `Parser`. This would basically bring it to agreement with how things are done for ex. in PDF Loaders.
- However I found out that I'd have to do this shuffling for a fairly long list of document loaders. Given how much time such effort would require, I opted to create `DocumentLoaderAsParser` wrapper. This wrapper can be used on any DocumentLoader that accepts `file_path` argument and turns it into `BaseBlobParser`. 

Hope this makes sense.